### PR TITLE
Closes #612: Switch from SB2.7.X to SB3.1.X and from javax to jakarta

### LINKED ISSUE #612: Upgrade to Spring Boot 3.1.X
Spring Boot 2.7.X and Spring Boot 3.0.X will end of life at the end of November. See the End of Life page: https://endoflife.date/spring-boot

Camunda upgraded the 7er line to Spring Boot 3.1.X already.

Taskana is close to being migrated to Spring Boot 3.1.x see https://github.com/Taskana/taskana/pull/2421

So we can and have to upgrade TaskanaAdapter.

Please try to update all dependencies to the latest version.
for example
this
```xml
    <dependency>
      <groupId>javax.xml.bind</groupId>
      <artifactId>jaxb-api</artifactId>
      <version>2.3.1</version>
    </dependency>
```
will be
```xml
    <dependency>
      <groupId>jakarta.xml.bind</groupId>
      <artifactId>jakarta.xml.bind-api</artifactId>
      <version>4.0.1</version>
    </dependency>
```

### PR DESCRIPTION
**Pull request by arolfes**
Sonar-Cloud: https://sonarcloud.io/summary/new_code?id=arolfes_TaskanaAdapter&branch=sb3_alex

Release Notes:

```markdown
# Breaking
* Move to Spring Boot 3.1.5
    * look for official migration guides of spring boot
        * Recommendations:
        * https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-3.0-Migration-Guide
        * https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-3.1-Release-Notes
        * https://www.baeldung.com/spring-boot-3-migration (not office but really recommended :) )
      * during the migration of TaskanaAdapter we faced some additional migration steps which are not covered in the guides:
        * in the `application.properties` you need to add 2 properties
          * `resteasy.jaxrs.app.registration=property`
          * `resteasy.jaxrs.app.classes=pro.taskana.adapter.camunda.outbox.rest.config.OutboxRestServiceConfig`
          * Without this config, the resteasy-servlet-spring-boot-starter will register the some TaskanaAdapterBeans twice, which does not work.
        * Camunda from version 7.20 on requires a default history time to live
          * `camunda.bpm.generic-properties.properties.historyTimeToLive: P180D`
          * see https://forum.camunda.io/t/camunda-7-20-history-time-to-live-cannot-be-null-cannot-deploy-wf-created-in-7-18/48159
* Switch to jakarta instead of javax wherever there is a jakarta equivalent.
* Switch to OKHTTPClient instead of JDK default
  * new parameters
    * `okhttp.connection-timeout` - time in milliseconds to establish a connection default 2000
    * `okhttp.read-timeout` - time in milliseconds to wait for response - default 5000
```

---
 
### For the submitter:
- [ ] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [ ] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [x] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
- [x] I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [ ] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [ ] I put the ticket in review

### Verified by the reviewer:
- [ ] Commit message format → ADPT-XXX: Your commit message.
- [ ] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [ ] SonarCloud analysis meets our standards
- [ ] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [ ] PR fulfills the ticket
- [ ] Edge cases and unwanted side effects are tested
- [ ] Readability

### CODE REVIEW

File: taskana-adapter-camunda-system-connector/src/main/java/pro/taskana/adapter/systemconnector/camunda/api/impl/HttpHeaderProvider.java [hunk 1]
```diff
@@ -43,8 +43,23 @@ public HttpHeaders outboxRestApiHeaders() {
     }
   }
 
-  HttpEntity<String> prepareEntityFromBodyForCamundaRestApi(String requestBody) {
+  public HttpEntity<Void> prepareEntityFromBodyForCamundaRestApi() {
```
Comment (ryzheboka) [inline taskana-adapter-camunda-system-connector/src/main/java/pro/taskana/adapter/systemconnector/camunda/api/impl/HttpHeaderProvider.java:46]: The "FromBody" part of the method name doesn't make sense to me.  The request body isn't used in this method anymore. What do you think about renaming this method? For example "newEntityWithCamundaRestApiHeaders"?

File: taskana-adapter-camunda-spring-boot-test/src/test/java/pro/taskana/adapter/integration/CamundaProcessengineRequester.java [hunk 2]
```diff
@@ -254,7 +264,7 @@ public boolean deleteProcessInstanceWithId(String processInstanceId, boolean ski
   public boolean isCorrectAssignee(String camundaTaskId, String assigneeValueToVerify) {
 
     String requestUrl = BASIC_ENGINE_PATH + this.processEngineKey + TASK_PATH + "/" + camundaTaskId;
-    HttpEntity<String> requestEntity = prepareEntityFromBody("{}");
+    HttpEntity<Void> requestEntity = httpHeaderProvider.prepareEntityFromBodyForCamundaRestApi();
```
Comment (ryzheboka) [inline taskana-adapter-camunda-spring-boot-test/src/test/java/pro/taskana/adapter/integration/CamundaProcessengineRequester.java:267]: Why do you change the method from "prepareEntityFromBody" to "prepareEntityFromBodyForCamundaRestAPI" so that it uses camunda header here?

File: taskana-adapter-camunda-system-connector/src/main/java/pro/taskana/adapter/systemconnector/camunda/config/CamundaSystemConnectorConfiguration.java [hunk 3]
```diff
@@ -16,10 +19,16 @@
...
 @DependsOn(value = {"adapterSpringContextProvider"})
+@EnableConfigurationProperties(OkHttpProperties.class)
 public class CamundaSystemConnectorConfiguration {
+
   @Bean
-  public RestTemplate restTemplate(RestTemplateBuilder builder) {
-    return builder.build();
+  RestTemplate restTemplate(RestTemplateBuilder builder, OkHttpProperties okHttpProperties) {
+    return builder
+        .setConnectTimeout(Duration.ofMillis(okHttpProperties.getConnectionTimeout()))
```
Comment (ryzheboka) [inline taskana-adapter-camunda-system-connector/src/main/java/pro/taskana/adapter/systemconnector/camunda/config/CamundaSystemConnectorConfiguration.java:28]: Why do you add configuration here?

File: taskana-adapter-camunda-system-connector/src/main/java/pro/taskana/adapter/systemconnector/camunda/config/OkHttpProperties.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package pro.taskana.adapter.systemconnector.camunda.config;
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@ConfigurationProperties(prefix = "okhttp")
+public class OkHttpProperties {
+
+  private long connectionTimeout = 2_000;
```
Comment (ryzheboka) [inline taskana-adapter-camunda-system-connector/src/main/java/pro/taskana/adapter/systemconnector/camunda/config/OkHttpProperties.java:10]: Why did you choose this values for the timeouts?

**Review by ryzheboka — CHANGES_REQUESTED** [review]

File: taskana-adapter-camunda-system-connector/src/main/java/pro/taskana/adapter/systemconnector/camunda/config/CamundaSystemConnectorConfiguration.java (hunk 3 above)
Comment (arolfes) [reply taskana-adapter-camunda-system-connector/src/main/java/pro/taskana/adapter/systemconnector/camunda/config/CamundaSystemConnectorConfiguration.java:28]: Hello @ryzheboka ,

maybe my comment in the commit message was not clear enough. I try to make it better this time. :)

Unfortunately, the default jdk client used so far does not support the HTTP request method PATCH. That's why I switched to the OkHttpClient.

You then need to tell the RestTemplateBuilder which HTTP client to use.

OkHttpClient comes with very large defaults for connection and read timeout. I don't know how to set these timeouts using the default Spring properties. That's why I introduced my own properties (OkHttpProperties) and set them.

I'm always a fan of allowing the user to set the timeouts themselves.

Does this answer your question?

File: taskana-adapter-camunda-system-connector/src/main/java/pro/taskana/adapter/systemconnector/camunda/config/OkHttpProperties.java (hunk 4 above)
Comment (arolfes) [reply taskana-adapter-camunda-system-connector/src/main/java/pro/taskana/adapter/systemconnector/camunda/config/OkHttpProperties.java:10]: Good question, there was no spec :)

The values come from a gut feeling.

Personally, I think 2 seconds to connect is more than enough. If it takes longer, you should have a look at your network or check where the request is getting stuck. (Normally the connection should be established within 500ms).

Well, and ReadTimeout is such a feeling. @gitgoodjhe can probably give you a better answer what a good default value is for reading data from Camunda.

Does it answer your question?

File: taskana-adapter-camunda-system-connector/src/main/java/pro/taskana/adapter/systemconnector/camunda/api/impl/HttpHeaderProvider.java (hunk 1 above)
Comment (arolfes) [reply taskana-adapter-camunda-system-connector/src/main/java/pro/taskana/adapter/systemconnector/camunda/api/impl/HttpHeaderProvider.java:46]: good point. I will change it.

Historically only these Methods exists

```Java

  HttpEntity<String> prepareEntityFromBodyForCamundaRestApi(String requestBody) {
    HttpHeaders headers = getHttpHeadersForCamundaRestApi();
    return new HttpEntity<>(requestBody, headers);
  }

  HttpHeaders getHttpHeadersForCamundaRestApi() {
    HttpHeaders headers = camundaRestApiHeaders();
    headers.setContentType(MediaType.APPLICATION_JSON);
    return headers;
  }

  HttpHeaders getHttpHeadersForOutboxRestApi() {
    HttpHeaders headers = outboxRestApiHeaders();
    headers.setContentType(MediaType.APPLICATION_JSON);
    return headers;
  }
```

but theses are not enough anymore. Thats why I created new methods without thinking about the names

File: taskana-adapter-camunda-spring-boot-test/src/test/java/pro/taskana/adapter/integration/CamundaProcessengineRequester.java (hunk 2 above)
Comment (arolfes) [reply taskana-adapter-camunda-spring-boot-test/src/test/java/pro/taskana/adapter/integration/CamundaProcessengineRequester.java:267]: Because I need also one for OutboxRestApi

File: taskana-adapter-camunda-outbox-rest-spring-boot-starter/pom.xml [hunk 5]
```diff
@@ -28,24 +28,16 @@
...
     <dependency>
       <groupId>org.postgresql</groupId>
       <artifactId>postgresql</artifactId>
       <scope>test</scope>
     </dependency>
-    <dependency>
-      <groupId>org.jboss.resteasy</groupId>
-      <artifactId>jaxrs-api</artifactId>
-      <version>${version.resteasy.jaxrs}</version>
-      <scope>provided</scope>
```
Comment (gitgoodjhe) [inline taskana-adapter-camunda-outbox-rest-spring-boot-starter/pom.xml:47]: why did you delete this?

**Review by gitgoodjhe — CHANGES_REQUESTED** [review]

File: taskana-adapter-camunda-spring-boot-test/src/main/resources/application.properties [hunk 6]
```diff
@@ -1,6 +1,10 @@
 server.port= 10020
 spring.main.allow-bean-definition-overriding=true
 
+# properties for resteasy-servlet-spring-boot-starter
+# without these 2 propertiers the camunda-context is registered twice
+resteasy.jaxrs.app.registration=property
```
[the remark this answers is not in the document]
Comment (arolfes) [reply taskana-adapter-camunda-spring-boot-test/src/main/resources/application.properties:6]: I see no tests :), Thats why I didn't stumbled about the problems :)
I will change it, add a test and then see how we need to adapt autoconfiguration

File: taskana-adapter-camunda-outbox-rest-spring-boot-starter/pom.xml (hunk 5 above)
Comment (arolfes) [reply taskana-adapter-camunda-outbox-rest-spring-boot-starter/pom.xml:47]: This replaced by the official jakarta-ws-rs-api

```xml
<dependency>
    <groupId>jakarta.ws.rs</groupId>
    <artifactId>jakarta.ws.rs-api</artifactId>
    <version>3.1.0</version>
</dependency>
```

File: taskana-adapter-camunda-spring-boot-test/src/main/resources/application.properties (hunk 6 above)
[the remark this answers is not in the document]
Comment (arolfes) [reply taskana-adapter-camunda-spring-boot-test/src/main/resources/application.properties:6]: Hi @gitgoodjhe , I've added the needed changes in a separate commit. Please review again

File: taskana-adapter-camunda-system-connector/src/main/java/pro/taskana/adapter/systemconnector/camunda/config/CamundaSystemConnectorConfiguration.java (hunk 3 above)
Comment (ryzheboka) [reply taskana-adapter-camunda-system-connector/src/main/java/pro/taskana/adapter/systemconnector/camunda/config/CamundaSystemConnectorConfiguration.java:28]: > OkHttpClient comes with very large defaults for connection and read timeout. I don't know how to set these timeouts using the default Spring properties. That's why I introduced my own properties (OkHttpProperties) and set them.

Thanks, that was the detail I needed :)

File: taskana-adapter-camunda-spring-boot-test/src/main/resources/application.properties (hunk 6 above)
[the remark this answers is not in the document]
Comment (arolfes) [reply taskana-adapter-camunda-spring-boot-test/src/main/resources/application.properties:6]: > In addition to that I'm wondering, why our SpringBoot Starter (Autoconfiguration) still seems to work. With SpringBoot 3 the spring.factories file isn't compatible anymore. I think we should definately change this to the new "org.springframework.boot.autoconfigure.AutoConfiguration.imports" file, what do you think? :)

Hi @gitgoodjhe, I've fixed it and added a testcase.

image

**Review by gitgoodjhe — APPROVED** [review]
